library/std/src/sys/net/connection/uefi/tcp4.rs

```diff
@@ -167,6 +168,31 @@ impl Tcp4 {
         }
     }
 
+    unsafe fn wait_or_cancel(
+        &self,
+        timeout: Option<Duration>,
+        event: r_efi::efi::Event,
+    ) -> io::Result<()> {
+        if !self.wait_for_flag(timeout) {
+            let _ = unsafe { self.cancel(event) };
+            return Err(io::Error::new(io::ErrorKind::TimedOut, "Operation Timed out"));
+        }
+
+        Ok(())
+    }
+
+    unsafe fn cancel(&self, event: r_efi::efi::Event) -> io::Result<()> {
+        let protocol = self.protocol.as_ptr();
+        let mut completion_token = tcp4::CompletionToken { event, status: Status::SUCCESS };
```

Comment:
From https://uefi.org/specs/UEFI/2.11/28_Network_Protocols_TCP_IP_and_Configuration.html#efi-ip4-protocol-cancel: "Pointer to a token that has been issued by EFI_IP4_PROTOCOL.Transmit() or EFI_IP4_PROTOCOL.Receive()". I read that to mean that the `completion_token` arg should be a pointer to the one created in the `read` and `write` methods above, not a new one.

(Also, if read literally it would seem like that means cancel can only be used for canceling receive/transmit, not connect. However, the [edk2 implementation](https://github.com/tianocore/edk2/blob/5ea0be305a072d995c11b163f6292d24cf514d0b/NetworkPkg/TcpDxe/SockInterface.c#L965) at least does not seem to have that restriction.)

Reply:
You are looking at wrong protocol. The correct one is here: https://uefi.org/specs/UEFI/2.11/28_Network_Protocols_TCP_IP_and_Configuration.html#efi-tcp4-protocol-cancel